### PR TITLE
Wallfollow intruder

### PR DESCRIPTION
**Pull request by Jolijn2020**
The basics are done

it can't deal with glass or teleports yet, but those are added as seperate issuess in GitHub

### CODE REVIEW

**Comment by Jolijn2020** [comment]
I goes directly towards the target until it hits a wall
follows the wall from the left or right (random rn) until it is on the line from beginPosition to targetPosition again
once on that line, it goes directly towards target again

File: src/main/java/app/model/Map.java [hunk 1]
```diff
@@ -78,7 +79,8 @@ public Map(Settings settings)
         {
             Vector srt = randPosition(intruderSpawn);
             Vector dir = randDirection();
-            Agent intruder = new WallFollowAgent(srt, dir, 10, Team.INTRUDER);
+            //Agent intruder = new WallFollowAgent(srt, dir, 10, Team.INTRUDER);
```
Comment (S010MON) [inline src/main/java/app/model/Map.java:82]: If this is being pulled into Master, remove commented out code

**Review by S010MON — CHANGES_REQUESTED** [review]
I think that @laurajogi should be requested to review as she knows this agent the best and would be the most useful reviewer

**Review by S010MON — COMMENTED** [review]
This seems to introduce two errors, I'm not sure if you realised, but as you are making changed to the graphics engine, you need to confirm that the system works for all agents, not just the one you modified.

1. I think that you have introduced a side effect to the other agent, running the simulation with guard agents causes a freeze with the following stack trace:

```
Exception in thread "JavaFX Application Thread" java.lang.RuntimeException: Cardinal point not found
	at app@1.0/app.model.agents.ACO.AcoAgent.detectCardinalPoint(AcoAgent.java:303)
	at app@1.0/app.model.agents.ACO.AcoAgent.explorationToViableMovement(AcoAgent.java:256)
	at app@1.0/app.model.agents.ACO.AcoAgent.visibleExploration(AcoAgent.java:247)
	at app@1.0/app.model.agents.ACO.AcoAgent.move(AcoAgent.java:56)
	at app@1.0/app.controller.GameEngine.tick(GameEngine.java:59)
	at app@1.0/app.controller.GameEngine.lambda$new$0(GameEngine.java:35)
	at javafx.graphics/com.sun.scenario.animation.shared.TimelineClipCore.visitKeyFrame(TimelineClipCore.java:239)
	at javafx.graphics/com.sun.scenario.animation.shared.TimelineClipCore.playTo(TimelineClipCore.java:180)
	at javafx.graphics/javafx.animation.Timeline.doPlayTo(Timeline.java:172)
	at javafx.graphics/javafx.animation.AnimationAccessorImpl.playTo(AnimationAccessorImpl.java:39)
	at javafx.graphics/com.sun.scenario.animation.shared.InfiniteClipEnvelope.timePulse(InfiniteClipEnvelope.java:104)
	at javafx.graphics/javafx.animation.Animation.doTimePulse(Animation.java:1189)
	at javafx.graphics/javafx.animation.Animation$1.lambda$timePulse$0(Animation.java:207)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at javafx.graphics/javafx.animation.Animation$1.timePulse(Animation.java:206)
	at javafx.graphics/com.sun.scenario.animation.AbstractPrimaryTimer.timePulseImpl(AbstractPrimaryTimer.java:343)
	at javafx.graphics/com.sun.scenario.animation.AbstractPrimaryTimer$MainLoop.run(AbstractPrimaryTimer.java:266)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:571)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:555)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.pulseFromQueue(QuantumToolkit.java:548)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.lambda$runToolkit$11(QuantumToolkit.java:353)
	at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
	at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
	at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$11(GtkApplication.java:290)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

https://user-images.githubusercontent.com/10490509/166136739-cb5f00a3-66cf-49d8-99f5-fb80fe941919.mp4



2.  Also, but less worryingly, the agents seem to get stuck on the diagonal quite frequently (see attached video).  This shouldn't stop your PR, but might be something to diagnose.

**Comment by S010MON** [comment]
@trav-d13 or @Travisd13  (I don't know which one is the imposter!) perhaps you can take a look at the changes that @Jolijn2020 made to the graphics engine, I think that it might be messing with the ACO agent's direction finding

**Comment by Jolijn2020** [comment]
I am aware of them getting stuck on the diagonal, there is made an issue for this.

It is cause by agents getting blocked by eachother, the current version just doesn't do anything and than repeats the move until it isn't blocked anymore. But since the agents sometimes move in opposite direction towards eachother they are both waiting for the other to move, causing a stale-mate.



As for the graphics engine, I was not aware of that, I'll check it out.

**Comment by Jolijn2020** [comment]
@S010MON @Travisd13 / @trav-d13 

I added 2 ACO-agent guards to the simulation, but did not get the same error. They did however, not move.

I also looked at your error messages, and saw that the origin of the error is related to the direction of the agent. The only related change I made that could influence the ACO agent is in the GameEngine, where I set the direction of the agent, to be the same as the direction in the Move-variable that the move() method returns.

Unless the ACO-agent doesn't return the correct direction in that Move-variable, it shouldn't cause any issues. I tried just putting the direction into the Move before returning it, but that didn't seem to really solve the issue. The ACO-agents are now moving, but extremely slow (almost didn't notice at first).

Commenting out the a.setDirection(move.getDirection) seemed to solve the issue though, so it is from that change. I'll look into how to fix it tomorrow morning.

If @Travisd13 / @trav-d13 has any idea when the direction has to be changed exactly, please let me know. Otherwise, I'll just look into it myself.

**Comment by Jolijn2020** [comment]
Okay, so looks like ACO-agent uses the EndDir-variable in the Move class, as a way to store the previous position.
So changing this messes up it's internal system.

Will probably have to create another variable that stores the position instead, so that EndDir can be used as the direction.

**Comment by Jolijn2020** [comment]
The AcoAgent now puts the direction into the Move-variable it returns, and stores the previousPosition in another (new) variable. So the GameEngine can now just set the direction for all agents

**Review by GwenTinho — APPROVED** [review]

**Review by S010MON — APPROVED** [review]
Looks good

**Pull request merged**